src/common.cuh

```diff
@@ -78,4 +78,105 @@ __host__ __device__ static void swap(T &a, T &b)
+
+
+typedef unsigned long long base_atomic_type;
+
+#ifdef COMPILE_FOR_CPU
+
+static base_atomic_type atomicCAS(base_atomic_type *address, const base_atomic_type compare, const base_atomic_type val)
+{
+    base_atomic_type ans = *address;
+
+    if(*address == compare)
+        *address = val;
+
+    return ans;
+}
+
+static base_atomic_type atomicAdd(base_atomic_type *address, const base_atomic_type value)
+{
+    base_atomic_type ans = *address;
+    *address += value;
+    return ans;
+}
+
+static double atomicAdd(double *address, const double value)
+{
+    double ans = *address;
+    *address += value;
+    return ans;
+}
+
+#endif //COMPILE_FOR_CPU
+
+// The following code is copied from https://stackoverflow.com/a/62094892/11248508 and modified (@kolayne can explain)
+// `address` CANNOT be pointer to const, because we are trying to edit memory by it's source
+__device__ static bool atomicCAS(bool *const address, const bool compare, const bool val)
+{
+    static_assert(sizeof(base_atomic_type) > 1, "The local atomicCAS implementation won't work if the size of "
+                                                "`base_atomic_type` is <= 1");
+    static_assert((sizeof(base_atomic_type) - 1 & sizeof(base_atomic_type)) == 0,
+                  "The local atomicCAS implementation won't work if the size of `base_atomic_type` is not a power of 2");
+
+
+    auto address_num = (unsigned long long)address;
+    unsigned pos = address_num & (sizeof(base_atomic_type) - 1);  // byte position within the `base_atomic_type`
+
+    auto *address_of_extended = (base_atomic_type *)(address - pos);  // `base_atomic_type`-aligned address
+    base_atomic_type old_extended = *address_of_extended, compare_extended, current_value_extended;
+
+    bool current_value;
+
+    do
+    {
+        current_value = (bool)(old_extended & ((0xFFU) << (8 * pos)));
+
+        if(current_value != compare) // If we expected that bool to be different, then
+            break; // stop trying to update it and just return it's current value
+
+        compare_extended = old_extended;
+
+        if(val)
+            current_value_extended = old_extended | (1U << (8 * pos));
+        else
+            current_value_extended = old_extended & (~((0xFFU) << (8 * pos)));
+
+        old_extended = atomicCAS(address_of_extended, compare_extended, current_value_extended);
+    } while(compare_extended != old_extended);
+
+    return current_value;
+}
+
+// I (Nikolay Nechaev, @kolayne) have no idea why the fuck the following only works with #else. If you're reading this
+// and now why, PLEASE, contact me and tell me
+#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 600
+#else
+// The following code is from https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomic-functions and
+// slightly changed (@kolayne can explain)
+__device__ static double atomicAdd(double* address, double val)
+{
+    auto* address_as_ull = (unsigned long long int*)address;
```

Comment:
Codacy found an issue: [Casting from double * to unsigned long long * is not portable due to different binary data representations on different platforms.](https://app.codacy.com/gh/physarumAdv/minds_crawl/pullRequest?prid=6278404)

Reply:
Hey, Codacy, wtf? This is a really old piece of code. Why did you only notice it today?

Reply:
This is a false positive. It would be correct if we were actually using the value at the `address` address as a ULL, but we are only going to work with this as with binary data (atomics do not care what is inside of the variables we are passing to them). I will ask Codacy to ignore this issue and recheck the PR